File: services/notifications.py
from datetime import datetime, timedelta

from fastapi import HTTPException
from sqlmodel import Session, select, col, func

from models.notification import Notification


def list_notifications(
    session: Session,
    skip: int = 0,
    limit: int = 50,
    is_read: bool | None = None,
    type_filter: str | None = None,
) -> list[Notification]:
    query = select(Notification).order_by(col(Notification.created_at).desc()).offset(skip).limit(limit)
    if is_read is not None:
        query = query.where(Notification.is_read == is_read)
    if type_filter:
        query = query.where(Notification.type == type_filter)
    return list(session.exec(query).all())


def get_notification(session: Session, notification_id: int) -> Notification:
    notification = session.get(Notification, notification_id)
    if not notification:
        raise HTTPException(status_code=404, detail="Notification not found")
    return notification


def create_notification(
    session: Session, type: str, title: str, body: str, related_entity: str | None = None
) -> Notification:
    notification = Notification(
        type=type, title=title, body=body, related_entity=related_entity
    )
    session.add(notification)
    session.commit()
    session.refresh(notification)
    return notification


def mark_read(session: Session, notification_id: int) -> Notification:
    notification = get_notification(session, notification_id)
    notification.is_read = True
    session.add(notification)
    session.commit()
    session.refresh(notification)
    return notification


def mark_all_read(session: Session) -> None:
    notifications = session.exec(
        select(Notification).where(Notification.is_read == False)  # noqa: E712
    ).all()
    for n in notifications:
        n.is_read = True
        session.add(n)
    session.commit()


def delete_notification(session: Session, notification_id: int) -> None:
    notification = get_notification(session, notification_id)
    session.delete(notification)
    session.commit()


def delete_all_read(session: Session) -> None:
    notifications = session.exec(
        select(Notification).where(Notification.is_read == True)  # noqa: E712
    ).all()
    for n in notifications:
        session.delete(n)
    session.commit()


def count_notifications(
    session: Session,
    is_read: bool | None = None,
    type_filter: str | None = None,
) -> int:
    query = select(func.count(Notification.id))
    if is_read is not None:
        query = query.where(Notification.is_read == is_read)
    if type_filter:
        query = query.where(Notification.type == type_filter)
    return int(session.exec(query).one())


def get_unread_count(session: Session) -> int:
    result = session.exec(
        select(func.count(Notification.id)).where(Notification.is_read == False)  # noqa: E712
    ).one()
    return int(result)


def cleanup_old_notifications(session: Session, max_age_days: int = 30, max_total: int = 500) -> int:
    """Delete old read notifications to prevent table bloat.

    Deletes read notifications older than max_age_days, and if the total count
    still exceeds max_total, trims the oldest read notifications down to that limit.
    Returns the number of deleted notifications.
    """
    deleted = 0

    # Phase 1: delete read notifications older than max_age_days
    cutoff = datetime.utcnow() - timedelta(days=max_age_days)
    old_read = session.exec(
        select(Notification).where(
            Notification.is_read == True,  # noqa: E712
            Notification.created_at < cutoff,
        )
    ).all()
    for n in old_read:
        session.delete(n)
        deleted += 1

    # Phase 2: if total still exceeds max_total, trim oldest read notifications
    total = session.exec(select(func.count(Notification.id))).one()
    if int(total) > max_total:
        excess = int(total) - max_total
        to_trim = session.exec(
            select(Notification)
            .where(Notification.is_read == True)  # noqa: E712
            .order_by(col(Notification.created_at))
            .limit(excess)
        ).all()
        for n in to_trim:
            session.delete(n)
            deleted += 1

    if deleted:
        session.commit()
    return deleted

File: yfine2/src/db/repo/notifications.ts
import type { SqlExecutor } from "../types";

export type NotificationType = "info" | "alert" | "warning";

export interface NotificationRow {
  id: number;
  type: NotificationType;
  title: string;
  body: string;
  related_entity: string | null;
  is_read: number;
  created_at: string;
}

const now = () => new Date().toISOString();

export async function createNotification(
  db: SqlExecutor,
  n: { type: NotificationType; title: string; body: string; related_entity?: string | null },
): Promise<void> {
  await db.execute(
    `INSERT INTO notifications (type,title,body,related_entity,is_read,created_at) VALUES (?,?,?,?,0,?)`,
    [n.type, n.title, n.body, n.related_entity ?? null, now()],
  );
}

/** True if an UNREAD notification exists for this exact related_entity. */
export async function hasUnread(db: SqlExecutor, relatedEntity: string): Promise<boolean> {
  const r = await db.select<{ c: number }>(
    `SELECT COUNT(*) c FROM notifications WHERE related_entity = ? AND is_read = 0`,
    [relatedEntity],
  );
  return (r[0]?.c ?? 0) > 0;
}

export async function listNotifications(
  db: SqlExecutor,
  opts: { limit?: number; unreadOnly?: boolean } = {},
): Promise<NotificationRow[]> {
  const where = opts.unreadOnly ? "WHERE is_read = 0" : "";
  return db.select<NotificationRow>(
    `SELECT id,type,title,body,related_entity,is_read,created_at FROM notifications ${where} ORDER BY created_at DESC, id DESC LIMIT ?`,
    [opts.limit ?? 100],
  );
}

export async function unreadCount(db: SqlExecutor): Promise<number> {
  const r = await db.select<{ c: number }>(`SELECT COUNT(*) c FROM notifications WHERE is_read = 0`);
  return r[0]?.c ?? 0;
}

export async function markRead(db: SqlExecutor, id: number): Promise<void> {
  await db.execute(`UPDATE notifications SET is_read = 1 WHERE id = ?`, [id]);
}

export async function markAllRead(db: SqlExecutor): Promise<void> {
  await db.execute(`UPDATE notifications SET is_read = 1 WHERE is_read = 0`);
}

export async function deleteNotification(db: SqlExecutor, id: number): Promise<void> {
  await db.execute(`DELETE FROM notifications WHERE id = ?`, [id]);
}
